cirq/google/serializable_device.py

```diff
@@ -0,0 +1,136 @@
+# Copyright 2019 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Device object for converting from device specification protos"""
+from typing import Any, cast, Dict, List, Set, Tuple, TYPE_CHECKING
+
+from cirq import devices, ops
+from cirq.google import serializable_gate_set
+from cirq.google.api.v2 import device_pb2
+from cirq.value import Duration
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class SerializableDevice(devices.Device):
+    """Device object generated from a device specification proto.
+
+    Given a device specification proto and a gate_set to translate the
+    serialized gate_ids to cirq Gates, this will generate a Device that can
+    verify operations and circuits for the hardware specified by the device.
+    """
+
+    def __init__(self, proto: device_pb2.DeviceSpecification,
```

Comment:
Regardless of the thread on the main discussion, I think you should have the constructor take the python types, i.e. qubits, allowed_targets, duration, target_sets and then factor out the protobuf stuff into a classmethod.

Among other benefits, this would let you document the class attributes which are currently undocumented. For example, I can't really tell what `self.target_sets` is going to be.

Reply:
Ok, let me think about the best way to do this.  This is still in progress.

Reply:
This has been done.